protocol/src/main/java/org/mifos/chatbot/protocol/telegram/TelegramChatService.java

```diff
@@ -0,0 +1,183 @@
+                if(user != null) {
+                    userRepository.removeUser(user.getUsername());
+                    sendTextMessage(senderId, "Logged out successfully.");
+                } else {
+                    sendTextMessage(senderId, "You are already logged out.");
+                }
+                return;
+            }
+            if (messageText.toLowerCase().contains("login:")) {
+                StringBuilder username = new StringBuilder();
+                StringBuilder password = new StringBuilder();
+                String creds = messageText.replaceAll("login:", "");
+                for (int i = 0; i < creds.length(); i++) {
+                    if (creds.charAt(i) == ':') {
+                        for (i++; i < creds.length(); i++) {
+                            password.append(creds.charAt(i));
+                        }
+                        break;
+                    }
+                    username.append(creds.charAt(i));
+                }
+                if(userRepository.telegramIDExist(Long.toString(senderId))) {
+                    sendTextMessage(senderId, "You are already logged in, please logout first.");
+                } else {
+                    if (authUser(username.toString(), password.toString())) {
+                        User user = userRepository.findUserByUsername(username.toString());
+                        if (user == null) {
+                            userRepository.addUserByTelegramID(
+                                    username.toString(),
+                                    password.toString(),
+                                    Long.toString(senderId)
+                            );
+                        } else {
+                            userRepository.updateUserTelegramID(
+                                    username.toString(),
+                                    password.toString(),
+                                    Long.toString(senderId)
+                            );
+                        }
+                        sendTextMessage(senderId, "Login successfully.");
+                    } else {
+                        sendTextMessage(senderId, "Please enter valid credentials.");
+                        return;
+                    }
+                }
+                return;
+            }
+            User user = userRepository.findUserByTelegramID(Long.toString(senderId));
+            if(user != null) {
+                String username = user.getUsername();
+                String password = user.getSecret_Pass();
+                if(authUser(username, password)) {
+                    ApiClient apiClient = new ApiClient(base64Encode(username + ":" + password));
+                    org.mifos.chatbot.client.Configuration.setDefaultApiClient(apiClient);
+                    List<MifosResponse> responseList = adapterService.handle(messageText.toLowerCase());
```

Comment:
This style is clean and clear. Good work!

Reply:
Thank you 😃